### PR TITLE
Feature monolith

### PR DESCRIPTION
**Pull request by pi-rho**
> puzzles: add in support for monolithic categories

This also supports static or dynamic puzzles within a category following the
earlier design strategy.

Tested on a monolithic category, needs testing with a variety of category types.

### CODE REVIEW

File: tools/moth.py [hunk 1]
```diff
@@ -328,15 +334,35 @@ def __init__(self, path, seed):
...
+            catmod = SourceFileLoader(
+                'catmod',
+                os.path.join(path, 'category.py')).load_module()
+            assert all([
+                hasattr(catmod, 'make'),
+                hasattr(catmod, 'points'),
+                type(catmod.points) is list,
+            ])
+            self.catmod = catmod
+            self.pointvals.extend(catmod.points)
```
Comment (nealey) [inline tools/moth.py]: If you define a category module, it needs to generate the entire category. This prevents the discrepancy where the category generates a point value that's also got a directory.

File: tools/moth.py [hunk 2]
```diff
@@ -328,15 +334,35 @@ def __init__(self, path, seed):
...
             points = int(pn)
             self.pointvals.append(points)
+
         self.pointvals.sort()
 
     def puzzle(self, points):
-        path = os.path.join(self.path, str(points))
-        return Puzzle(self.seed, path)
+        if self.catmod is not None and points in self.catmod.points:
+            return Puzzle(self.seed, points=points, category=self.catmod)
```
Comment (nealey) [inline tools/moth.py]: This feels like a cumbersome design; I'd like puzzles to not need to know what category they're in. Consider:

```python
if self.catmod:
    p = Puzzle(self.seed, points=points)
    catmod.make(p, points)
```

File: tools/moth.py (hunk 1 above)
Comment (pi-rho) [inline tools/moth.py]: As stated in the commit message, this follows the precedent set earlier -- if individual puzzles are dynamic, they don't all have to be dynamic; also, a puzzle may be both dynamic and static. Category follows this same pattern. Corollary: if it's not good for Category, Puzzle needs to change as well. :smile:

File: tools/moth.py (hunk 2 above)
Comment (pi-rho) [inline tools/moth.py]: toe-may-toe/toe-mah-toe ... The initializer for Puzzle is fairly strict and needs a signal that it is to be dynamically made. Passing the category-maker instance is the signal. We could revert to a PuzzleFactory pattern and make Puzzle a smart dict with no instantiation as well.

File: tools/moth.py (hunk 1 above)
Comment (pflarr) [inline tools/moth.py]: I kind of agree with Patrick on this one, and it's fairly easy to handle.  

It's not so much that puzzle handles all these different cases and chooses one, instead it handles all of them in a well defined order.

File: tools/moth.py (hunk 2 above)
Comment (pflarr) [inline tools/moth.py]: I've proposed a newer, differenter pull request.  See that.

File: tools/moth.py (hunk 1 above)
Comment (nealey) [inline tools/moth.py]: Nope. The presence of `category.py` means you're generating the category. The category author doesn't get to then drop in other things piecemeal and cause us to have to sort out how to deal with there being two 20-point puzzles.

Note that this doesn't prevent the author from calling `puzzle.read_directory()` in their make function.

**Pull request merged**